File: ios/DailyMemory/Domain/UseCase/Photo/AnalyzeImageUseCase.swift
import Foundation
import UIKit

/// Use case for analyzing images with AI
/// Extracts objects, scenes, text, and generates tags
final class AnalyzeImageUseCase {
    private let imageAnalysisService: ImageAnalysisService

    init(imageAnalysisService: ImageAnalysisService = .shared) {
        self.imageAnalysisService = imageAnalysisService
    }

    /// Analyze a single image and return results
    func execute(image: UIImage) async -> Result<PhotoAnalysis, Error> {
        let result = await imageAnalysisService.analyzeImage(image)

        switch result {
        case .success(let analysisResult):
            return .success(analysisResult.toPhotoAnalysis())
        case .failure(let error):
            return .failure(error)
        }
    }

    /// Analyze image from file URL
    func execute(imageURL: URL) async -> Result<PhotoAnalysis, Error> {
        let result = await imageAnalysisService.analyzeImage(from: imageURL)

        switch result {
        case .success(let analysisResult):
            return .success(analysisResult.toPhotoAnalysis())
        case .failure(let error):
            return .failure(error)
        }
    }

    /// Analyze multiple images and combine results
    func execute(images: [UIImage]) async -> Result<[PhotoAnalysis], Error> {
        var results: [PhotoAnalysis] = []

        for image in images {
            let result = await imageAnalysisService.analyzeImage(image)
            switch result {
            case .success(let analysisResult):
                results.append(analysisResult.toPhotoAnalysis())
            case .failure(let error):
                return .failure(error)
            }
        }

        return .success(results)
    }

    /// Analyze images and merge into combined analysis
    func executeAndMerge(images: [UIImage]) async -> Result<CombinedPhotoAnalysis, Error> {
        let results = await execute(images: images)

        switch results {
        case .success(let analyses):
            return .success(CombinedPhotoAnalysis.merge(analyses))
        case .failure(let error):
            return .failure(error)
        }
    }
}

// MARK: - Photo Analysis Model

/// Analysis result for a single photo
struct PhotoAnalysis: Equatable {
    let objects: [String]
    let scene: String
    let ocrText: String?
    let faceCount: Int
    let description: String
    let suggestedTags: [String]

    /// Check if people were detected
    var hasPeople: Bool {
        faceCount > 0 || objects.contains { $0.lowercased().contains("person") }
    }

    /// Check if text was detected
    var hasText: Bool {
        ocrText != nil && !ocrText!.isEmpty
    }
}

/// Combined analysis from multiple photos
struct CombinedPhotoAnalysis: Equatable {
    let allObjects: [String]
    let scenes: [String]
    let allOcrText: [String]
    let totalFaces: Int
    let descriptions: [String]
    let mergedTags: [String]

    /// Merge multiple photo analyses into one
    static func merge(_ analyses: [PhotoAnalysis]) -> CombinedPhotoAnalysis {
        var allObjects: Set<String> = []
        var scenes: [String] = []
        var allOcrText: [String] = []
        var totalFaces = 0
        var descriptions: [String] = []
        var allTags: Set<String> = []

        for analysis in analyses {
            allObjects.formUnion(analysis.objects)
            scenes.append(analysis.scene)
            if let text = analysis.ocrText, !text.isEmpty {
                allOcrText.append(text)
            }
            totalFaces += analysis.faceCount
            descriptions.append(analysis.description)
            allTags.formUnion(analysis.suggestedTags)
        }

        return CombinedPhotoAnalysis(
            allObjects: Array(allObjects).sorted(),
            scenes: scenes,
            allOcrText: allOcrText,
            totalFaces: totalFaces,
            descriptions: descriptions,
            mergedTags: Array(allTags).sorted()
        )
    }
}

// MARK: - Extension

extension ImageAnalysisResult {
    func toPhotoAnalysis() -> PhotoAnalysis {
        PhotoAnalysis(
            objects: objects,
            scene: scene,
            ocrText: text,
            faceCount: faces,
            description: description,
            suggestedTags: suggestedTags
        )
    }
}

File: android/app/src/main/java/com/dailymemory/domain/usecase/photo/AnalyzeImageUseCase.kt
package com.dailymemory.domain.usecase.photo

import android.graphics.Bitmap
import android.net.Uri
import com.dailymemory.data.remote.ImageAnalysisResult
import com.dailymemory.data.remote.ImageAnalysisService
import java.io.File
import javax.inject.Inject

/**
 * Use case for analyzing images with AI
 * Extracts objects, scenes, text, and generates tags
 */
class AnalyzeImageUseCase @Inject constructor(
    private val imageAnalysisService: ImageAnalysisService
) {
    /**
     * Analyze a single bitmap and return results
     */
    suspend operator fun invoke(bitmap: Bitmap): Result<PhotoAnalysis> {
        return imageAnalysisService.analyzeImage(bitmap).map { result ->
            result.toPhotoAnalysis()
        }
    }

    /**
     * Analyze image from URI
     */
    suspend fun analyze(uri: Uri): Result<PhotoAnalysis> {
        return imageAnalysisService.analyzeImage(uri).map { result ->
            result.toPhotoAnalysis()
        }
    }

    /**
     * Analyze image from file
     */
    suspend fun analyze(file: File): Result<PhotoAnalysis> {
        return imageAnalysisService.analyzeImage(file).map { result ->
            result.toPhotoAnalysis()
        }
    }

    /**
     * Analyze multiple bitmaps and return individual results
     */
    suspend fun analyzeMultiple(bitmaps: List<Bitmap>): Result<List<PhotoAnalysis>> {
        val results = mutableListOf<PhotoAnalysis>()

        for (bitmap in bitmaps) {
            val result = imageAnalysisService.analyzeImage(bitmap)
            result.onSuccess { analysisResult ->
                results.add(analysisResult.toPhotoAnalysis())
            }.onFailure { error ->
                return Result.failure(error)
            }
        }

        return Result.success(results)
    }

    /**
     * Analyze multiple bitmaps and merge into combined analysis
     */
    suspend fun analyzeAndMerge(bitmaps: List<Bitmap>): Result<CombinedPhotoAnalysis> {
        return analyzeMultiple(bitmaps).map { analyses ->
            CombinedPhotoAnalysis.merge(analyses)
        }
    }
}

/**
 * Analysis result for a single photo
 */
data class PhotoAnalysis(
    val objects: List<String>,
    val scene: String,
    val ocrText: String?,
    val faceCount: Int,
    val description: String,
    val suggestedTags: List<String>
) {
    /**
     * Check if people were detected
     */
    val hasPeople: Boolean
        get() = faceCount > 0 || objects.any { it.lowercase().contains("person") }

    /**
     * Check if text was detected
     */
    val hasText: Boolean
        get() = !ocrText.isNullOrEmpty()
}

/**
 * Combined analysis from multiple photos
 */
data class CombinedPhotoAnalysis(
    val allObjects: List<String>,
    val scenes: List<String>,
    val allOcrText: List<String>,
    val totalFaces: Int,
    val descriptions: List<String>,
    val mergedTags: List<String>
) {
    companion object {
        /**
         * Merge multiple photo analyses into one
         */
        fun merge(analyses: List<PhotoAnalysis>): CombinedPhotoAnalysis {
            val allObjects = mutableSetOf<String>()
            val scenes = mutableListOf<String>()
            val allOcrText = mutableListOf<String>()
            var totalFaces = 0
            val descriptions = mutableListOf<String>()
            val allTags = mutableSetOf<String>()

            for (analysis in analyses) {
                allObjects.addAll(analysis.objects)
                scenes.add(analysis.scene)
                analysis.ocrText?.takeIf { it.isNotEmpty() }?.let { allOcrText.add(it) }
                totalFaces += analysis.faceCount
                descriptions.add(analysis.description)
                allTags.addAll(analysis.suggestedTags)
            }

            return CombinedPhotoAnalysis(
                allObjects = allObjects.toList().sorted(),
                scenes = scenes,
                allOcrText = allOcrText,
                totalFaces = totalFaces,
                descriptions = descriptions,
                mergedTags = allTags.toList().sorted()
            )
        }
    }
}

/**
 * Extension to convert API result to domain model
 */
private fun ImageAnalysisResult.toPhotoAnalysis(): PhotoAnalysis {
    return PhotoAnalysis(
        objects = objects,
        scene = scene,
        ocrText = text,
        faceCount = faces,
        description = description,
        suggestedTags = suggestedTags
    )
}
